practices/appeng/adrs/0002-distributed-tracing.md

```diff
@@ -0,0 +1,140 @@
+- works across multiple applications, programming languages, and transports
+- improve time to market
+- facilitates excellent cross-team communication and cooperation
+
+Here are several example scenarios or questions that distributed tracing can
+help answer.
+
+- As a new engineer on the team, I want to understand how many separate systems
+  are involved when a certain user type logs in and the first page is rendered.
+- As an operations engineer, I want to know how many SQL queries are executed
+  for a given endpoint or interaction.
+- As a product manager, I want to know if a new feature is being used by a
+  certain cohort of users on a regular basis.
+- As an engineer, I want to prove that an optimization I wrote is effective
+  in a production environment.
+- As a load tester, after I have shown that a problem exists, I want to
+  understand how the system is interacting so I can debug and fix the issue.
+
+### ADR Goals and Anti-goals
+
+- Goal: Choose which set of libraries to use at callsites (across programming
+  languages) within the codebase, which will be used to generate
+  distributed tracing data
+- Anti-goal: Committing to a specific "backend", i.e. platform or service for
+  gathering, exploring, and displaying trace information
+  - Sub-goal: Leaving open as many options as possible for the backend
+
+## Decision
+
+## Use OpenTelemetry
+
+- OpenTelemetry is an emerging industry standard
+  - vendors find benefit of being in the OpenTelemetry ecosystem because they
+    no longer have to create or support instrumentation libraries in an ever
+    growing array of languages, i.e. as soon as language library exists for
+    OpenTelemetry, the vendors automatically become available to support that
+    given language.
+- OpenTelemetry is vendor agnostic
+  - tracing information can be sent to hosted services (e.g. Honeycomb.io, AWS
+    X-Ray, etc) or self-hosted Open Source implementations (e.g. Zipkin, Jaeger,
+    etc)
+  - if left unconfigured, OpenTelemetry instrumentation calls default to
+    lightweight/noop executions
+- OpenTelemetry has well-maintained libraries for the languages used in the
+  layers of most Truss projects
+  - i.e. Go (back-end); JavaScript (front-end); Python (load testing); etc
+- Easily swappable back-ends
+  - e.g. could choose a local Docker version of OpenZipkin for an all-local
+    development environment
+  - e.g. can use Honeycomb.io in the experimental commercial-cloud hosted
+    environment
+  - e.g. can swap in AWS X-Ray for use in GovCloud hosted environments
+- Cons
+  - as an abstraction layer, OpenTelemetry may prohibit usage of vendor-
+    specific capabilities
+  - some OpenTelemetry libraries and tools may trail their vendor-supported
+    counterparts
+  - instrumentation for tracing may be a vector for performance overhead
+
+## Why is this Applicable to the Practice as a Whole
+
+TODO: What makes this decision generally applicable
```

Comment:
Most applications will benefit from instrumentation once they are accepting production traffic, and for many systems, the amount of work required to add tracing calls to the code is non-trivial and needs to be done preemptively.

Projects should identify what types of calls need to be traced as a part of discovery and document these decisions in a project-level ADR. For example, a web backend might add tracing to HTTP requests, service invocations, database calls, and any network requests made by the backend.

Reply:
LOL same brain!

I added some thoughts on this in the [other comment here](https://github.com/trussworks/Engineering-Playbook/pull/205#discussion_r741496096) but this definitely seems like the more appropriate space for it. 

Given this, I'd say that it might be interesting (currently) to assume that we shouldn't assume that OpenTelemetry or application tracing is something we would suggest against given how costly it can be to do it in the future.

Considering the future of web applications, it seems like Observability is something that is here to stay. Maybe this means we need to consider that in this ADR.